### DISCUSSION TITLE
Reading a specific page from a multi-page TIFF

### DISCUSSION THREAD

**Discussion by MrAnderson1971** [ImageMagick/ImageMagick#6274] [Help] [answered]
### ImageMagick version

7.1.1-6-Q16

### Operating system, version and so on

Windows 10 (x86)

### Description

Is it possible to read a specific page of a multi-page TIFF from a memory buffer using the C++ api? 

`Magick::Blob blob(buffer, buffer_size);`

`Magick::Image image(blob);`
only reads the first page

**Comment by snibgo** [comment]
> only reads the first page

Does it? I don't do much with blobs, but as far as I can see, if your buffer contains more than one image, your returned `Image` will be a list of images. (Technically, you always get a list of images, but the list might have only one image.)

Try writing your `Image` to `x.tiff`, and see what you get.

**Reply by MrAnderson1971** [reply]
Yeah I tried it on a multi-page TIFF and only got one page when I reopened it with GIMP. The `read` description does say it reads a "single image frame"

That being said, if it is a list of images, how can I choose which one image to write?

**Comment by snibgo** [comment, marked as answer]
I suggest you post a minimal but complete program, from reading a multi-image tiff to writing another, like this MagickCore example:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <MagickCore/MagickCore.h>

int main(int argc, char **argv) {
  ExceptionInfo
    *exception;

  Image
    *images;

  ImageInfo
    *image_info;

  MagickCoreGenesis(*argv, MagickTrue);

  MagickInfo *entry;

  exception = AcquireExceptionInfo();
  image_info = CloneImageInfo((ImageInfo *)NULL);

 /* Assume we have done:
      magick rose: wizard: x.tiff
 */

  size_t length = 322102;
  char *buffer = calloc(length, sizeof(char *));
  FILE *file = fopen ("x.tiff", "r");
  fread (buffer, length, sizeof(char), file);
  images = BlobToImage (image_info, buffer, length, exception);

  if (exception->severity != UndefinedException) {
    printf("%s\n", exception->reason);
    exit(1);
  }

  if (images == (Image *)NULL)
    exit(1);

  (void)strcpy (images->filename, "output.tiff");
  CopyMagickString (image_info->magick,"TIFF",MagickPathExtent);
  CopyMagickString (image_info->magick,"",MagickPathExtent);
  WriteImage (image_info, images, exception);
  return (0);
}
```
When the program is built and run, we examine output.tiff:
```
magick output.tiff info:

output.tiff[0] TIFF 70x46 70x46+0+0 8-bit TrueColor sRGB 322102B 0.000u 0:00.000
output.tiff[1] TIFF 480x640 480x640+0+0 8-bit Palette sRGB 256c 0.000u 0:00.001
```

> That being said, if it is a list of images, how can I choose which one image to write?

When you have variable `image` of type `Image`, `image->next` is the next in the list, or `NULL` if there are no more images.

**Reply by MrAnderson1971** [reply]
Thank you.

**Comment by dlemstra** [comment]
If you only want to read a specific page you should set `subImage` of the `image` before reading from the blob.
